File: SenLib/Sen3/FileFixes/c0830_dat.cs
using HyoutaUtils;
using HyoutaUtils.Streams;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace SenLib.Sen3.FileFixes {
	class c0830_dat : FileMod {
		public string GetDescription() {
			return "Terminology fixes in Crossbell (Heiyue).";
		}

		public IEnumerable<FileModResult> TryApply(FileStorage storage) {
			var file = storage.TryGetDuplicate(new HyoutaUtils.Checksum.SHA1(0xbb6a5b0a31ebe05cul, 0xeb06ae8c0b773e2ful, 0x0b8823afu));
			if (file == null) {
				return null;
			}

			var bin = file.CopyToMemory();

			// Heiyue Trading Company -> Heiyue trading company
			bin.Position = 0x19bc;
			bin.WriteUInt8(0x74);
			bin.Position = 0x19c4;
			bin.WriteUInt8(0x63);
			bin.Position = 0x685e;
			bin.WriteUInt8(0x74);
			bin.Position = 0x6866;
			bin.WriteUInt8(0x63);

			return new FileModResult[] { new FileModResult("data/scripts/scena/dat_en/c0830.dat", bin) };
		}

		public IEnumerable<FileModResult> TryRevert(FileStorage storage) {
			var file = storage.TryGetDuplicate(new HyoutaUtils.Checksum.SHA1(0xbb6a5b0a31ebe05cul, 0xeb06ae8c0b773e2ful, 0x0b8823afu));
			if (file == null) {
				return null;
			}
			return new FileModResult[] { new FileModResult("data/scripts/scena/dat_en/c0830.dat", file) };
		}
	}
}

File: native/sen3/FileFixes/c0830_dat.cpp
#include <string_view>
#include <vector>

#include "sen/file_getter.h"
#include "p3a/pack.h"
#include "p3a/structs.h"
#include "sha1.h"

namespace SenLib::Sen3::FileFixes::c0830_dat {
std::string_view GetDescription() {
    return "Terminology fixes in Crossbell (Heiyue).";
}

bool TryApply(const SenPatcher::GetCheckedFileCallback& getCheckedFile,
              std::vector<SenPatcher::P3APackFile>& result) {
    try {
        auto file = getCheckedFile(
            "data/scripts/scena/dat_en/c0830.dat",
            30265,
            SenPatcher::SHA1FromHexString("bb6a5b0a31ebe05ceb06ae8c0b773e2f0b8823af"));
        if (!file) {
            return false;
        }

        auto& bin = file->Data;

        // Heiyue Trading Company -> Heiyue trading company
        bin[0x19bc] = 0x74;
        bin[0x19c4] = 0x63;
        bin[0x685e] = 0x74;
        bin[0x6866] = 0x63;

        result.emplace_back(std::move(bin), file->Filename, SenPatcher::P3ACompressionType::LZ4);

        return true;
    } catch (...) {
        return false;
    }
}
} // namespace SenLib::Sen3::FileFixes::c0830_dat
